File: pages/stories/the_lighthouse/update_summaries.js
// Node.js script to update summaries to clean prose format
const fs = require('fs');
const path = require('path');

const jsonPath = path.join(__dirname, 'RobertEggers_2019_TheLighthouse_scenes_analyzed.json');

function parseScreenplayText(text) {
    const lines = text.split('\n');
    const actionDescriptions = [];
    const dialogueExchanges = [];
    let currentSpeaker = null;
    let currentDialogue = [];

    for (let line of lines) {
        line = line.trim();
        if (!line) continue;

        // Skip page numbers like "77."
        if (/^\d+\.$/.test(line)) continue;

        // Check if it's a character name (all caps, short)
        if (/^[A-Z][A-Z\s']+(\s*\(.*\))?$/.test(line) && line.length < 30) {
            if (currentSpeaker && currentDialogue.length > 0) {
                dialogueExchanges.push([currentSpeaker, currentDialogue.join(' ')]);
            }
            currentSpeaker = line.split('(')[0].trim();
            currentDialogue = [];
        } else if (currentSpeaker) {
            // Skip stage directions in parentheses
            if (/^\(.*\)$/.test(line)) continue;
            if (line === '--') {
                currentDialogue.push('[pause]');
            } else {
                currentDialogue.push(line);
            }
        } else {
            actionDescriptions.push(line);
        }
    }

    if (currentSpeaker && currentDialogue.length > 0) {
        dialogueExchanges.push([currentSpeaker, currentDialogue.join(' ')]);
    }

    return { actionDescriptions, dialogueExchanges };
}

function generateCleanSummary(scene) {
    const text = scene.text || '';
    if (!text) return 'No description available.';

    const { actionDescriptions, dialogueExchanges } = parseScreenplayText(text);
    const summaryParts = [];

    // Action/Visual description
    if (actionDescriptions.length > 0) {
        let actionText = actionDescriptions.join(' ');
        // Remove page numbers and screenplay directions
        actionText = actionText.replace(/\b\d+\.\s*/g, '');
        actionText = actionText.replace(/\s+/g, ' ').trim();
        actionText = actionText.replace(/\b(CLOSE ON|WIDE|EXTREMELY WIDE SHOT|CUT TO|FADE|HOLD):\s*/g, '');
        actionText = actionText.replace(/\bHold\.\s*/g, '');
        if (actionText) {
            summaryParts.push(actionText);
        }
    }

    // Dialogue
    if (dialogueExchanges.length > 0) {
        const dialogueNarrative = [];
        for (const [speaker, dialogue] of dialogueExchanges) {
            let cleanDialogue = dialogue.replace(/\b\d+\.\s*/g, '').replace(/\s+/g, ' ').trim();
            if (cleanDialogue && cleanDialogue !== '[pause]') {
                if (cleanDialogue.length > 200) {
                    cleanDialogue = cleanDialogue.substring(0, 197) + '...';
                }
                dialogueNarrative.push(`${speaker}: "${cleanDialogue}"`);
            } else if (cleanDialogue === '[pause]') {
                dialogueNarrative.push(`${speaker} remains silent.`);
            }
        }
        if (dialogueNarrative.length > 0) {
            summaryParts.push(dialogueNarrative.join(' '));
        }
    }

    let fullSummary = summaryParts.join(' ').replace(/\s+/g, ' ').trim();

    // Fallback if too short
    if (fullSummary.length < 50 && text) {
        let cleanText = text.replace(/\b\d+\.\s*/g, '').replace(/\s+/g, ' ').trim();
        if (cleanText.length > 600) {
            cleanText = cleanText.substring(0, 597) + '...';
        }
        fullSummary = cleanText;
    }

    return fullSummary;
}

// Main
console.log('Loading JSON...');
const data = JSON.parse(fs.readFileSync(jsonPath, 'utf-8'));

console.log(`Processing ${data.scenes.length} scenes...`);
for (const scene of data.scenes) {
    const summary = generateCleanSummary(scene);
    if (!scene.plotSummary) scene.plotSummary = {};
    scene.plotSummary.detailed = summary;
    scene.summary = summary;
}

console.log('Writing updated JSON...');
fs.writeFileSync(jsonPath, JSON.stringify(data, null, 2), 'utf-8');

console.log('Done! Sample summaries:');
for (const i of [1, 23, 98]) {
    if (i <= data.scenes.length) {
        const summary = data.scenes[i - 1].summary;
        console.log(`\n--- Scene ${i} ---`);
        console.log(summary.length > 300 ? summary.substring(0, 300) + '...' : summary);
    }
}

File: pages/stories/the_lighthouse/update_summaries.py
#!/usr/bin/env python3
"""
Generate clean prose summaries for each scene in The Lighthouse.
No markdown formatting - just readable text.
"""

import json
import re

def parse_screenplay_text(text):
    """Parse screenplay text into action descriptions and dialogue exchanges."""
    lines = text.split('\n')

    action_descriptions = []
    dialogue_exchanges = []
    current_speaker = None
    current_dialogue = []

    for line in lines:
        line = line.strip()
        if not line:
            continue

        # Skip page numbers like "77." or "78."
        if re.match(r'^\d+\.$', line):
            continue

        # Check if it's a character name (all caps, typically short)
        if re.match(r'^[A-Z][A-Z\s\']+(\s*\(.*\))?$', line) and len(line) < 30:
            # Save previous dialogue if any
            if current_speaker and current_dialogue:
                dialogue_exchanges.append((current_speaker, ' '.join(current_dialogue)))
            # Start new speaker
            current_speaker = line.split('(')[0].strip()
            current_dialogue = []
        elif current_speaker:
            # Skip stage directions in parentheses like "(suddenly)"
            if re.match(r'^\(.*\)$', line):
                continue
            # It's dialogue
            if line == '--':
                current_dialogue.append('[pause]')
            else:
                current_dialogue.append(line)
        else:
            # It's action/description (no current speaker)
            action_descriptions.append(line)

    # Capture last dialogue
    if current_speaker and current_dialogue:
        dialogue_exchanges.append((current_speaker, ' '.join(current_dialogue)))

    return action_descriptions, dialogue_exchanges


def generate_clean_summary(scene):
    """Generate a clean prose summary without markdown."""
    text = scene.get('text', '')

    if not text:
        return "No description available."

    # Parse the screenplay text
    action_descriptions, dialogue_exchanges = parse_screenplay_text(text)

    summary_parts = []

    # Action/Visual description - main narrative
    if action_descriptions:
        action_text = ' '.join(action_descriptions)
        action_text = re.sub(r'\b\d+\.\s*', '', action_text)  # Remove page numbers
        action_text = re.sub(r'\s+', ' ', action_text).strip()
        if action_text:
            summary_parts.append(action_text)

    # Dialogue summary - convert to readable format
    if dialogue_exchanges:
        dialogue_narrative = []
        for speaker, dialogue in dialogue_exchanges:
            dialogue = re.sub(r'\b\d+\.\s*', '', dialogue)
            dialogue = re.sub(r'\s+', ' ', dialogue).strip()

            if dialogue and dialogue != '[pause]':
                if len(dialogue) > 250:
                    dialogue = dialogue[:247] + "..."
                dialogue_narrative.append(f'{speaker}: "{dialogue}"')
            elif dialogue == '[pause]':
                dialogue_narrative.append(f'{speaker} remains silent.')

        if dialogue_narrative:
            summary_parts.append(' '.join(dialogue_narrative))

    # Combine into prose
    full_summary = ' '.join(summary_parts)
    full_summary = re.sub(r'\s+', ' ', full_summary).strip()

    # Fallback if too short
    if len(full_summary) < 50 and text:
        clean_text = re.sub(r'\b\d+\.\s*', '', text)
        clean_text = re.sub(r'\s+', ' ', clean_text).strip()
        if len(clean_text) > 800:
            clean_text = clean_text[:797] + "..."
        full_summary = clean_text

    return full_summary


def main():
    print("Loading JSON...")
    with open('RobertEggers_2019_TheLighthouse_scenes_analyzed.json', 'r', encoding='utf-8') as f:
        data = json.load(f)

    print(f"Processing {len(data['scenes'])} scenes...")

    for scene in data['scenes']:
        detailed_summary = generate_clean_summary(scene)

        if 'plotSummary' not in scene:
            scene['plotSummary'] = {}
        scene['plotSummary']['detailed'] = detailed_summary
        scene['summary'] = detailed_summary

    print("Writing updated JSON...")
    with open('RobertEggers_2019_TheLighthouse_scenes_analyzed.json', 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

    print("Done! Sample summaries:")
    for i in [1, 23, 98]:
        if i <= len(data['scenes']):
            summary = data['scenes'][i-1]['summary']
            print(f"\n--- Scene {i} ---")
            print(summary[:400] + "..." if len(summary) > 400 else summary)


if __name__ == '__main__':
    main()

